engine/server/engine/streaming/websocket_pump.go

```diff
@@ -0,0 +1,119 @@
+package streaming
+
+import (
+	"context"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/labstack/echo/v4"
+	"github.com/rs/cors"
+	"github.com/sirupsen/logrus"
+
+	api_type "github.com/kurtosis-tech/kurtosis/api/golang/http_rest/api_types"
+)
+
+const (
+	wsReadBufferSize  = 1024
+	wsWriteBufferSize = 1024
+	maxMessageSize    = 512
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 60 * time.Second
+	// Send pings to peer with this period. Must be less than pongWait.
+	// nolint:gomnd
+	pingPeriod = (pongWait * 9) / 10
+)
+
+type WebsocketPump[T interface{}] struct {
+	websocket  *websocket.Conn
+	inputChan  chan *T
+	infoChan   chan *api_type.ResponseInfo
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+}
+
+func NewWebsocketPump[T interface{}](ctx echo.Context, cors cors.Cors) (*WebsocketPump[T], error) {
+	// nolint: exhaustruct
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  wsReadBufferSize,
+		WriteBufferSize: wsWriteBufferSize,
+		CheckOrigin:     cors.OriginAllowed,
+	}
+
+	conn, err := upgrader.Upgrade(ctx.Response(), ctx.Request(), nil)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Failed to upgrade http connection to websocket")
+	}
+
+	ctxWithCancel, cancelFunc := context.WithCancel(context.Background())
+
+	return &WebsocketPump[T]{
+		websocket:  conn,
+		inputChan:  make(chan *T),
+		ctx:        ctxWithCancel,
+		cancelFunc: cancelFunc,
+	}, nil
+}
+
+func (pump WebsocketPump[T]) StartPumping() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		pump.websocket.Close()
+	}()
+
+	logrus.WithFields(logrus.Fields{
+		"pongWait":       pongWait,
+		"pingPeriod":     pingPeriod,
+		"maxMessageSize": maxMessageSize,
+	}).Debug("Started keep alive process for websocket connection.")
+
+	pump.websocket.SetReadLimit(maxMessageSize)
+	if err := pump.websocket.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		logrus.WithError(err).Error("Failed to set Pong wait time")
```

Comment:
will the client receive and error if this happens?

Reply:
no, this is logged only the server side. if this fails, it's very likely the connection is broken.